presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3ClientFactory.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.s3;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.Protocol;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.metrics.RequestMetricCollector;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Builder;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.facebook.presto.hive.HiveClientConfig;
+import io.airlift.units.Duration;
+import org.apache.hadoop.conf.Configuration;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
+import static com.amazonaws.regions.Regions.US_EAST_1;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_ENDPOINT;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_PIN_CLIENT_TO_CURRENT_REGION;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+
+/**
+ * This factory provides AmazonS3 client required for executing S3SelectPushdown requests.
+ * Normal S3 GET requests use AmazonS3 clients initialized in PrestoS3FileSystem or EMRFS.
+ * The ideal state will be to merge this logic with the two file systems and get rid of this
```

Comment:
Can you elaborate on this? What would the integration look like? What are the blockers to doing this now?

Reply:
It appears there is duplication here with `PrestoS3FileSystem`. Can we eliminate this duplication (even without merging the S3 Select code into the file system)?

Reply:
>Can you elaborate on this? What would the integration look like? What are the blockers to doing this now?

Sure will do

> It appears there is duplication here with PrestoS3FileSystem. Can we eliminate this duplication (even without merging the S3 Select code into the file system)?

Presto currently uses two filesystems, EMRFS and PrestoS3FileSystem. The ideal state for this code is to be integrated with both filesystems such that we won't need this. If we remove any logic from this class and merge it with PrestoS3FileSystem then support on EMR Presto integration will have regressions. I believe that after this initial push we can have dedicated effort to bring the two filesystems in sync, where an engineer with access to both code bases will make the required changes. We will need some planning to orchestrate the roll out so as to not cause any customer impact.

Reply:
Thanks, that makes sense. Until that happens, is it worth removing the duplication for the client configuration/setup between this class and `PrestoS3FileSystem`? Or we think this will happen soon and then this new class can be removed?

Reply:
> Until that happens, is it worth removing the duplication for the client configuration/setup between this class and PrestoS3FileSystem?

For routing traffic to S3 Select Presto uses the singleton client provided by this factory. The client is initialized by Guice during service startup. If certain parts of the query can't be pushed down then they are routed to the underlying file system implementation for service. The problem with removing the code from the factory would be that the singleton client initialization on EMR would fail (or get partially initialized) for customers using EMRFS. Other dimension of the problem is that the code is already released in production on EMR starting EMR release label 5.18. We are continuously applying fixes identified in the PR with every release going forward. Thus, we will still need some co-ordination with EMR team to prepare a patch for EMR for the initialization. 

>Or we think this will happen soon and then this new class can be removed?
We don't have an ETA for the required refactor at this point. However, we already have had this discussion, and both EMR and S3 Select team are aware of this duplication of code. I think we will need help from Presto community in order to prioritize effort on integration test, new file format support and code refactor. How about we create a set of issues for the feature on Github, and rank them in the order of priority so that our teams can start addressing them from top down.